src/security-rules/security-rules-api-client.ts

```diff
@@ -0,0 +1,92 @@
+/*!
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { HttpRequestConfig, HttpClient, HttpError } from '../utils/api-request';
+import { PrefixedFirebaseError } from '../utils/error';
+import { FirebaseSecurityRulesError, SecurityRulesErrorCode } from './security-rules-utils';
+
+const RULES_API_URL = 'https://firebaserules.googleapis.com/v1/';
+
+/**
+ * Class that facilitates sending requests to the Firebase security rules backend API.
+ *
+ * @private
+ */
+export class SecurityRulesApiClient {
+
+  constructor(private readonly httpClient: HttpClient) { }
+
+  /**
+   * Gets the specified resource from the rules API. Resource name must be full qualified names with project
+   * ID prefix (e.g. `projects/project-id/rulesets/ruleset-name`).
+   *
+   * @param {string} name Full qualified name of the resource to get.
+   * @returns {Promise<T>} A promise that fulfills with the resource.
+   */
+  public getResource<T>(name: string): Promise<T> {
+    if (!name.startsWith('projects/')) {
```

Comment:
may be worth regexing this to catch malformed project ids (which, for instance, I think need to be lowercase).

Reply:
I simplified things a bit here by pushing the projectId to the constructor of the `RulesApiClient`. Now it stores a partial URL as part of its state:

```
this.url = `${RULES_API_URL}/projects/${projectId}`;
```

Callers should pass in the resource names without project ID:

```
client.getResource('rulesets/my-ruleset');
```

As for project ID validation, we are doing the same check here as our other modules (Auth, FCM), and so it should be sufficient.

Reply:
I'll defer to you here, but:

- there aren't that many kinds of resources
- it may be easier to explicitly spell them all out as unique methods:
```
client.getRuleset(<ruleset_name>)
client.listRulesets()
client.getRelease()
client.listReleases()
```
which directly encodes the exposed backend API. Then add another layer on top of that.

Reply:
This is a good idea. Will implement in the next PRs.